[REVIEW] LIST Support for ColumnVector [skip-ci]

kuhushukla opened the pull request:
This is an _early_ WIP change to show how list type column vectors would work on host and device side. I have more asserts and cleanup work (extra lines and nits and consolidating methods - getting rid of some assumptions) to do but I wanted to put it out there to see how we can improve the host side design specifically and what other semantic changes would be required.

GPUtester commented:
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

kuhushukla commented:
There is a lot of cleanup and hardcoded stuff I need to fix along with of course more tests, so pardon that, especially on the host side. Mostly looking for design change comments.

revans2 commented on java/src/main/java/ai/rapids/cudf/ColumnViewPointerAccess.java:
```diff
@@ -0,0 +1,39 @@
+
+package ai.rapids.cudf;
+
+public interface ColumnViewPointerAccess {
```
Can we get some javadocs on this to explain what it is for?

kuhushukla replied:
Will get this in the next round here soon.

revans2 commented on java/src/main/java/ai/rapids/cudf/HostColumnVector.java:
```diff
@@ -21,10 +21,16 @@
 import org.slf4j.LoggerFactory;
 
-import java.nio.charset.StandardCharsets;
+import java.io.*;
```
I generally prefer not to have .* imports unless there are a huge number of them.

kuhushukla replied:
Agree, intellij barfed on me with imports.

revans2 commented on java/src/main/java/ai/rapids/cudf/HostColumnVector.java:
```diff
@@ -44,7 +50,9 @@
 
   private final OffHeapState offHeap;
   private final DType type;
+  private List<DType> allTypes = new ArrayList<>();
```
This is rather confusing without some documentation about what it is and how it is used.  Also how is this supposed to work with structs?  I know that is not a requirement right now, but we should be thinking about it, because it is coming very soon.

kuhushukla replied:
Thanks Bobby. Yes, it need docs for sure and in my experience this is the harder part of how we want the host side to look like. Without complicating these types further, I think having HostColumnVector deviate from ColumnVector and having some sort of recursive structure may be the way forward to keep it readable and making it work for structs.
For example: 
```HostColumnVector {
childHostColumnVectorMeta[] childMetas;
}
```
 This idea is similar the initial recursive approach we spoke offline about with  a `childHostColumnVectorMeta[]` per `ColumnVector` but we should certainly iron this out as part of this change.

revans2 commented on java/src/main/java/ai/rapids/cudf/HostColumnVector.java:
```diff
@@ -44,7 +50,9 @@
   private final DType type;
+  private List<DType> allTypes = new ArrayList<>();
   private long rows;
+  private List<Long> allRows = new ArrayList<>();
```
Same comment here like for `allTypes`

revans2 commented on java/src/main/java/ai/rapids/cudf/HostColumnVector.java:
```diff
@@ -75,14 +83,42 @@
       throw new IllegalStateException("Buffer cannot have a nullCount without a validity buffer");
     }
-    if (type != DType.STRING) {
+    if (type != DType.STRING && type != DType.LIST) {
```
Should we extend DType to include metadata about offsets?  This probably does not matter too much.

revans2 commented on java/src/main/java/ai/rapids/cudf/HostColumnVector.java:
```diff
@@ -75,14 +83,42 @@
+      throw new IllegalStateException("Buffer cannot have a nullCount without a validity buffer");
+    }
+    if (type.get(0) != DType.LIST) {
+      assert offsetBuffer == null : "offsets are only supported for STRING";
```
Should this be LIST instead of STRING?

revans2 commented on java/src/main/java/ai/rapids/cudf/HostColumnVector.java:
```diff
@@ -201,6 +237,24 @@ public boolean hasNulls() {
     return getNullCount() > 0;
   }
 
+    void printOffsetBuffer(int level, int prevSize) {
```
These debug commands should be removed.

revans2 commented on java/src/main/java/ai/rapids/cudf/HostColumnVector.java:
```diff
@@ -201,6 +237,24 @@ public boolean hasNulls() {
+  }
+
+
+    void printDataBuffer() {
```
Here too.

revans2 commented on java/src/test/java/ai/rapids/cudf/ColumnVectorTest.java:
```diff
@@ -2412,4 +2416,344 @@ void testCreateDurationNanoseconds() {
+
+  @Test
+  void testConcatListsStrings() {
+    List<String> list = new ArrayList<>();
```
There are much cleaner ways to create a list.

```
static import java.utils.Arrays.asList;

...
try ( ColumnVector res1 = ColumnVector.fromLists(DType.STRING,
                                         asList("0", "1", "2", "3"),
                                         null);
        ColumnVector res2 = ColumnVector.fromLists(DType.STRING,
                                         asList("4", null, "6", null));
...
```

kuhushukla replied:
Agree. Will get this in the next round.

revans2 commented:
I think a lot of the code that we have to build up a nested column and to copy it needs to be recursive so we can extend them to support structs in the future. To make some of the recursive processing work, though, we are going to need to do some of the tricks that you are doing with `stealColumnView`, which I really don't like.  I would very much prefer to have a separate purpose build class for this type of situation.

The other things I would like to see more of it thought about how to build up a column.  There really are two situations that we care about.

1) test code.  This has nothing to do with performance and has everything to do with readability.
2) production code. This has everything to do with performance and if the code is a bit cumbersome it is okay.

For the test code we have been able to get away with building an entire column at a time because even if the physical layout was the same for multiple types, we could tag it with the proper metadata by using a different function call, or passing in the DType directly. This is not going to work well with nested types, especially with structs.  It becomes even harder when we need to support decimal and a DType is not going to be enough. We need to think what we want the test code to look like so we can make it happen.  I think I would like to see something like the following.  Note that this can happen as separate follow on work.

```
// Create a list of list of strings column
column(listType(listType(stringType()),
    listData(listData("1", "2", "3"), null), // [["1", "2", "3"], null]
    listData()) // <- empty list
column(structType(stringType(), decimalType(5, 10), byteType()),
    structData("FIRST", "10.0", 5), // For testing it would be nice to have the test data try to make what was input fit with the types we want from the schema, but would need to do bounds checking etc.
    structData("SECOND", 100.2, "1")); 
```

For production data we are either going to be copying the data columnar to columnar or row to columnar. I think the simplest way to make this work is to use a builder pattern, like we have already used, but with sub-builders for child columns.
```
builder = columnBuilder(listType(nullable = false, size = 2, growable=false,
                                                    listType(nullable = true, size = 3, growable = false,
                                                                 stringType(nullable = false, bufferSize = 3, size = 3, bufferGrowable = true))));

firstList = builder.getChild(0);
secondList = firstList.getChild(0);
secondList.appendUTF8StringEntry("1".getBytes())
secondList.appendUTF8StringEntry("2".getBytes())
secondList.appendUTF8StringEntry("3".getBytes())
// Not sure if we want explicit synchronization points between the different levels or not.
firstList.appendNull();
builder.appendEmptyList();
builder.build();

//Or 
secondList.appendRawBytesWiothOffsets("123".getBytes(), new int [] {0, 1, 2, 3})
// Not sure if we want explicit synchronization points between the different levels or not.
firstList.appendNull();
builder.appendEmptyList();
builder.build();
```

kuhushukla commented:
Thanks @revans2  for the comments. Really helpful.

I am going to focus on for the next update here:
>  I would very much prefer to have a separate purpose build class for this type of situation.
>
and
> think the simplest way to make this work is to use a builder pattern, like we have already used, but with sub-builders for child columns.
>

kuhushukla commented:
> I would very much prefer to have a separate purpose build class for this type of situation. 
>
d135a0173d1804e218e707c35578bb892d9085b9 should address this.

kuhushukla commented:
@revans2 I am going to make changes to HostColumnVector and the builder to have something prototyped for you to take a look at based on your suggestion. I think we need to rework it as part of this PR to make testing work.

kuhushukla commented:
@revans2  I have made HCV recursive (`cleanImpl` might need to be done still), added a new prototype builder for lists (of int and strings only for now), added `DataType` for `ListType` etc. for test-ability and improved `getList()` call. Let me know what you think.

revans2 commented on java/src/main/java/ai/rapids/cudf/ColumnVector.java:
```diff
@@ -265,32 +354,81 @@ public HostColumnVector copyToHost() {
       HostMemoryBuffer hostDataBuffer = null;
```
Can we please make `copyToHost` call a recursive version of itself.  This is getting really confusing, and conceptually it should be very simple. I used `HostColumnVectorStandIn` here because conceptually the `copyToDevice` code should look very similar and I just wanted to show conceptually what it would look like. Here we might be able to get away with just a modified `HostColumnVector` or a `NestedHostColumnVector`.

```
private static final HostColumnVectorStandIn copyToHost(long deviceColumnView) {
  // Now some of this we could speed up with caching, but conceptually it should be like the following...
  // None of these need to be closed because they do not own any data, just views conceptually.
  DeviceMemoryBufferView offset = getOffsetView(deviceColumnView);
  DeviceMemoryBufferView valid = getValidView(deviceColumnView);
  DeviceMemoryBufferView data = getDataView(deviceColumnView);
  HostMemoryBuffer hostOffset = null;
  HostMemoryBuffer hostValid = null;
  HostMemoryBuffer hostData = null;
  List<HostColmnVectorStandIn> children = new List<>();
  try {
    for (long childView: getChildren(deviceColumnView)) { // Don't include offset children, etc
      children.add(copyToHost(childView));
    }
    // copy offset, valid and data to host as needed
    // Ownership of children and host buffers is transferred to new HostColumnVectorStandIn
    return new HostColumnVectorStandIn(hostOffset, hostValid, hostData, children);
  } finally {
    // close children if needed
    // close host buffers if needed
  }
}

public HostColumnVector copyToHost() {
  return copyToHost(getColumnView()).asHostColumnVector();
}
```

kuhushukla replied:
>Can we please make copyToHost call a recursive version of itself.
>
Yes that is my intent, however since the previous version was all iterative, I am doing the rework onepiece at a time. There are a couple more places where I rely on the older lists of offsets, validity etc. WIll fix in the next round of commits.

kuhushukla replied:
I have made `copytToHost` and `copyTodevice` recursive and also removed getBuffersToClose on the ColumnVector specifically. The recursive changes to these methods are close to what u imagined I think @revans2 but I could use some more input on this if u feel we need to further tighten the ownership when we do `createNestedColumnVector` and such.

revans2 commented on java/src/main/java/ai/rapids/cudf/ColumnVector.java:
```diff
@@ -2622,6 +2743,46 @@ private static native long rollingWindow(long viewHandle, int min_periods, int a
+
+  @Override
+  public MemoryBuffer getValidityBuffer() {
+    //TODO: fix me
```
What is wrong here?

kuhushukla replied:
Dead code, will remove

revans2 commented on java/src/main/java/ai/rapids/cudf/ColumnVector.java:
```diff
@@ -2845,21 +3050,147 @@ protected boolean cleanImpl(boolean logErrorIfNotClean) {
+    return NestedColumnVector.createColumnVector(data, valid, offsets, allTypes, allRows, nullCount);
+  }
+
+  private static class NestedColumnVector {
```
Can we just make `copyToHost` and `copyToDevice` closer to what I put up above in the comments? I think you have the recursion here at the wrong level.

kuhushukla replied:
Done.

revans2 commented on java/src/main/java/ai/rapids/cudf/HostColumnVector.java:
```diff
@@ -42,11 +49,148 @@
+      return ret;
+    }
+
+    public Object getElement(int rowIndex) throws IOException {
```
This is horribly inefficient. We need a different API for exporting data.

kuhushukla replied:
@revans2 , do u have any suggestions on how to pull back to host efficiently for a list?

revans2 replied:
The main thing is to put comments on the public APIs that are horribly bad.  these are ones that are going to create objects which means GC issues. This API creates lots of objects.

Doing something efficient would require more knowledge about memory layout so that we can copy large blocks of data at once.  This we can do with Unsafe and copy the data to an array, which is probably the preferred option if we want to do something performant.  But for now, just put in the comments and we can have follow on work to expose fast APIs.

kuhushukla replied:
Can you point me to the java code that might be doing something analogous to allow columnar to row to work in a performant way? We can keep these getters internal for testing and add something that would be more performant if you share some pointers. Thanks!

revans2 replied:
Columnar to row is going to require lots of investigation and performance tuning.  Please don't try to boil the ocean here.  Lets just put the comment on it and file a follow on issue.

kuhushukla replied:
Sounds good. I was not sure if it was a Must-have for this PR. Thanks!

revans2 commented on java/src/main/java/ai/rapids/cudf/HostColumnVector.java:
```diff
@@ -264,10 +467,65 @@ public ColumnVector copyToDevice() {
   // DATA ACCESS
   /////////////////////////////////////////////////////////////////////////////
 
+  public List getList(long rowIndex) throws Exception {
```
If we are going to expose this we need lots and lots of comments to say that this is very slow and should never be used in production.

kuhushukla commented on java/src/main/java/ai/rapids/cudf/ColumnVector.java:
```diff
@@ -24,18 +24,86 @@
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
+import javax.xml.datatype.DatatypeConfigurationException;
```
There is cleanup I need to do. Agree this is not needed.

kuhushukla commented on java/src/main/java/ai/rapids/cudf/ColumnVector.java:
```diff
@@ -24,18 +24,86 @@
+    }
+  }
+
+  private List<MemoryBuffer> getToClose() {
```
@revans2 , are we ok with passing the `NestedColumnVector` to the `ColumnVector` constructor so that it can then pick out the additional buffers it needs to track and the viewhandle (which is passed in right now to allow `makeCudfColumnView`) ?

mythrocks commented on java/src/main/native/src/ColumnVectorJni.cpp:
```diff
@@ -1097,15 +1106,44 @@ JNIEXPORT jlong JNICALL Java_ai_rapids_cudf_ColumnVector_makeCudfColumnView(
   CATCH_STD(env, 0);
 }
 
+JNIEXPORT jlong JNICALL Java_ai_rapids_cudf_ColumnVector_getChildColumnView(JNIEnv *env,
```
Is the intention here to get to the `column_view` of the list's child?

kuhushukla replied:
yes

kuhushukla replied:
I dont think it will be very useful in this patch but I kept it to see if this is worth keeping.

I use `getChildCvPointer` to do the child view pointer work on java side.

mythrocks replied:
I wonder if the following would be equivalent:
```c++
JNIEXPORT jlong JNICALL Java_ai_rapids_cudf_ColumnVector_getChildColumnView(
  JNIEnv *env,
  jobject j_object, jlong handle) {
  
  using namespace cudf;

  jni::auto_set_device(env);

  auto col {reinterpret_cast<column_view*>(handle)};
  return col->type().id() == type_id::LIST?
      new column_view(lists_column_view(col).child()) : 0;
}
```

The latest list code attempts to keep the child vectors intact, even for empty lists.
Also, I suspect that returning a default-constructed `column_view` would not be in keeping with the protocol of returning `nullptr` for invalid columns.

mythrocks replied:
I realize that this might be moot: `getChildColumnView()` should really take a child column index.

kuhushukla replied:
Good point! Let me see what I can do here.

kuhushukla replied:
Done.

kuhushukla commented on java/src/main/java/ai/rapids/cudf/ColumnVector.java:
```diff
@@ -24,18 +24,86 @@
+    }
+
+    @Override
+    public long getRows() {
```
U got it.

kuhushukla replied:
done.

kuhushukla commented on java/src/main/java/ai/rapids/cudf/ColumnVector.java:
```diff
@@ -24,18 +24,86 @@
+    }
+
+    @Override
+    public ColumnViewPointerAccess getChildColumnView() {
```
Yes, that is needed. Will address and comment

mythrocks replied:
That might be beyond the scope of the current PR. We'll do that when we extend it for `struct`.

kuhushukla replied:
Sure, let me see if I can do this in the next set of commits , this should be easy.

kuhushukla replied:
@mythrocks can we resolve this now? I have addressed the concern here.

kuhushukla commented on java/src/main/java/ai/rapids/cudf/ColumnVector.java:
```diff
@@ -84,7 +151,24 @@ public ColumnVector(DType type, long rows, Optional<Long> nullCount,
+                      DeviceMemoryBuffer dataBuffer, DeviceMemoryBuffer validityBuffer,
+                      DeviceMemoryBuffer offsetBuffer, long childCv) {
+    if (type != DType.STRING && type != DType.LIST) {
+      assert offsetBuffer == null : "offsets are only supported for STRING";
```
Thanks, yes lot of cleanup needs to be done. As for the print buffers, I , like you, do find them useful.

kuhushukla commented on java/src/main/java/ai/rapids/cudf/ColumnVector.java:
```diff
@@ -2845,21 +3073,181 @@ protected boolean cleanImpl(boolean logErrorIfNotClean) {
+    }
+
+    private static NestedColumnVector createNestedColumnVector(DType type, int rows, Optional<Long> nullCount,
+                                                               HostMemoryBuffer dataBuffer, HostMemoryBuffer validityBuffer,
```
Name this better please

kuhushukla commented on java/src/main/java/ai/rapids/cudf/DeviceMemoryBuffer.java:
```diff
@@ -53,13 +53,13 @@ protected boolean cleanImpl(boolean logErrorIfNotClean) {
       }
       if (neededCleanup && logErrorIfNotClean) {
-        log.error("A DEVICE BUFFER WAS LEAKED (ID: " + id + " " + Long.toHexString(origAddress) + ")");
+        log.error("A DEVICE BUFFER WAS LEAKED (ID: " + lengthInBytes + " " + origAddress + ")");
```
Remove this change

kuhushukla commented on java/src/main/java/ai/rapids/cudf/ColumnVector.java:
```diff
@@ -24,18 +24,75 @@
+    @Override
+    public void close() {
+      //TODO TBD
+    }
```
should be removed I suppose

revans2 replied:
Yes

kuhushukla replied:
Doing that now.

kuhushukla commented on java/src/main/java/ai/rapids/cudf/ColumnVector.java:
```diff
@@ -257,6 +331,56 @@ public boolean hasNulls() {
 
+  private final HostColumnVector.NestedHostColumnVector copyToHostNestedHelper(ColumnViewPointerAccess deviceCvPointer) {
+    // Now some of this we could speed up with caching, but conceptually it should be like the following...
+    // None of these need to be closed because they do not own any data, just views conceptually.
```
Remove copy paste comments

revans2 commented:
I just tried to do the following 
```
           ColumnVector expected = ColumnVector.fromLists(
                   new HostColumnVector.ColumnBuilder.ListType(true, 6,
                           new HostColumnVector.ColumnBuilder.BasicType(true, 8, DType.STRING)),
                   Arrays.asList("Héllo", "there"),
                   Arrays.asList("thésé"),
                   null,
                   Arrays.asList(),
                   Arrays.asList("ARé", "some"),
                   Arrays.asList("test", "strings"));
```

and I ended up with getting an out of memory error. If I don't try to insert a null into the base column everything works fine.

kuhushukla commented:
> I just tried to do the following
> 
> ```
>            ColumnVector expected = ColumnVector.fromLists(
>                    new HostColumnVector.ColumnBuilder.ListType(true, 6,
>                            new HostColumnVector.ColumnBuilder.BasicType(true, 8, DType.STRING)),
>                    Arrays.asList("Héllo", "there"),
>                    Arrays.asList("thésé"),
>                    null,
>                    Arrays.asList(),
>                    Arrays.asList("ARé", "some"),
>                    Arrays.asList("test", "strings"));
> ```
> 
> and I ended up with getting an out of memory error. If I don't try to insert a null into the base column everything works fine.
Taking a look. Nulls did give me some grief recently it seems like I may have missed a case for nulls some place. Thanks for the catch.

kuhushukla commented:
> > I just tried to do the following
> > ```
> >            ColumnVector expected = ColumnVector.fromLists(
> >                    new HostColumnVector.ColumnBuilder.ListType(true, 6,
> >                            new HostColumnVector.ColumnBuilder.BasicType(true, 8, DType.STRING)),
> >                    Arrays.asList("Héllo", "there"),
> >                    Arrays.asList("thésé"),
> >                    null,
> >                    Arrays.asList(),
> >                    Arrays.asList("ARé", "some"),
> >                    Arrays.asList("test", "strings"));
> > ```
> > 
> > 
> > and I ended up with getting an out of memory error. If I don't try to insert a null into the base column everything works fine.
> 
> Taking a look. Nulls did give me some grief recently it seems like I may have missed a case for nulls some place. Thanks for the catch.
>

Latest commits add this test case even if temporarily and fixes the corner case of `asList()` by doing this specifically during `append()`

```
        if (inputList.isEmpty()) {
          if (childBuilder.type == DType.STRING) {
            childBuilder.append("");
          }
        }
```

kuhushukla commented:
Going to test empty list plus list with empty strings more as I think we need to set the offsets with more deliberation than the current code.

kuhushukla commented:
I would think in
```
ColumnVector expected = ColumnVector.fromLists(
                   new HostColumnVector.ColumnBuilder.ListType(true, 6,
                           new HostColumnVector.ColumnBuilder.BasicType(true, 8, DType.STRING)),
                   Arrays.asList("Héllo", "there"),
                   Arrays.asList("thésé"),
                   null,
                   Arrays.asList(),
                   Arrays.asList("ARé", "some"),
                   Arrays.asList("test", "strings"));

```
Arrays.asList() should increment rowcount only on the list column and not on the string column which has 7 valid rows. I would like to get an idea how this works semantically for lists.

kuhushukla commented:
@mythrocks, @razajafri  , do the latest changes specifically on the `ColumnViewPointerAccess` side look like something that is more palatable for structs? Let me know and I can refine it as we need. I have also made it so that data sits where it belongs for the most part.

I have only a couple hard codes left that we can change when structs go in. (mostly in copyToHost and such).

revans2 commented on java/src/main/java/ai/rapids/cudf/ColumnVector.java:
```diff
@@ -257,6 +348,58 @@ public boolean hasNulls() {
   // DATA MOVEMENT
   /////////////////////////////////////////////////////////////////////////////
 
+  private final HostColumnVector.NestedHostColumnVector copyToHostNestedHelper(ColumnViewPointerAccess deviceCvPointer) {
```
I think this can be static.

kuhushukla replied:
Done and fixed nullCount in the process here.

revans2 commented on java/src/main/java/ai/rapids/cudf/ColumnVector.java:
```diff
@@ -257,6 +348,58 @@ public boolean hasNulls() {
+      currValidity = deviceCvPointer.getValidityBuffer();
+      if (currData != null) {
+        hostData = HostMemoryBuffer.allocate(currData.length);
+        hostData.copyFromDeviceBuffer((DeviceMemoryBufferView)currData);
```
nit: Can we have a follow on issue to remove the casting?  We should be able to use a sub-interface or sub-class of `ColumnViewPointerAccess` that returns `DeviceMemoryBufferView` instead.  That way we can guarantee that the cast will work by way of the compiler.  Don't need to do it now we are in burn-down and this is a private method, so I would much rather just get this in as is.

kuhushukla replied:
Templatized to make the casting go away.

revans2 commented on java/src/main/java/ai/rapids/cudf/ColumnVector.java:
```diff
@@ -265,33 +408,78 @@ public HostColumnVector copyToHost() {
+          HostColumnVector ret = new HostColumnVector(type, rows, nullCount,
+              hData, hValid, hOffset, children);
+          return ret;
+          //TODO: close h buffers
```
This still needs to be done

kuhushukla replied:
was a lingering todo.

revans2 commented on java/src/main/java/ai/rapids/cudf/ColumnVector.java:
```diff
@@ -265,33 +408,78 @@ public HostColumnVector copyToHost() {
-      BaseDeviceMemoryBuffer valid = offHeap.getValid();
-      BaseDeviceMemoryBuffer offsets = offHeap.getOffsets();
-      BaseDeviceMemoryBuffer data = offHeap.getData();
+      BaseDeviceMemoryBuffer valid = (BaseDeviceMemoryBuffer)getValidityBuffer();
```
Another nit for follow on work. After refactoring the types being returned we probably could combine the majority of this code with the code from `copyToHostNestedHelper` so that all this does is call into `copyToHostNestedHelper` and then has an API to translate the top level `HostColumnVector.NestedHostColumnVector` into a `HostColumnVector`.  But again lets get this in as is and worry about that later.

kuhushukla replied:
Templatized the interface to get rid of the casting.

kuhushukla replied:
Sorry wrong reply here. I will do this as follow on.

revans2 replied:
Leaving this comment so we can capture the follow on work.

kuhushukla replied:
I did add the needed to change to take this out.

revans2 commented on java/src/main/java/ai/rapids/cudf/ColumnVector.java:
```diff
@@ -257,6 +348,58 @@ public boolean hasNulls() {
+      return new HostColumnVector.NestedHostColumnVector(currType, currRows, nullCount, hostData,
+          hostValid, hostOffsets, children);
+    } finally {
+      if (currData != null) {
```
The host buffers need to be closed on error here too, just like with the TODO down below in the original `copyToHost`

kuhushukla replied:
done.

revans2 commented on java/src/main/java/ai/rapids/cudf/ColumnVector.java:
```diff
@@ -265,33 +408,78 @@ public HostColumnVector copyToHost() {
+      Long rows = this.rows;
+      // hardcoded for lists for now
+      ColumnViewPointerAccess leafChildWithData = getChildColumnView(0);
+      while (leafChildWithData != null && leafChildWithData.getChildColumnView(0) != null) {
```
Can we have an explanation as to why we need to go two levels deep if there are two levels, but only one if there is just one?

kuhushukla replied:
Added some info let me know if that makes sense.

revans2 replied:
I still don't understand why data is different. I understand that we are treating offsets differently and so we have to play around with the number of children and offsets into the children, but data should just be in the leaf column right? If we play the same game that we did for Strings don't we risk losing the validity for that leaf column when we do this? If that is what we are doing here then I would argue that eventually we should just make string match list.

revans2 commented on java/src/main/java/ai/rapids/cudf/ColumnVector.java:
```diff
@@ -2815,19 +3091,20 @@ protected boolean cleanImpl(boolean logErrorIfNotClean) {
         neededCleanup = true;
       }
-      if (data != null || valid != null || offsets != null) {
+      //TODO: use toClose data structure
```
I think this is done.

revans2 commented on java/src/main/java/ai/rapids/cudf/ColumnVector.java:
```diff
@@ -2845,21 +3122,181 @@ protected boolean cleanImpl(boolean logErrorIfNotClean) {
+      this.child = child;
+    }
+
+    static ColumnVector createColumnVector(DType type, int rows, HostMemoryBuffer data, HostMemoryBuffer valid, HostMemoryBuffer offsets,
```
Can we get some java docs on these.  It is not really clear why this API is hear instead of a part of `ColumnVector` itself.  It is also not 100% clear what this API does.

revans2 commented on java/src/main/java/ai/rapids/cudf/ColumnVector.java:
```diff
@@ -2845,21 +3122,181 @@ protected boolean cleanImpl(boolean logErrorIfNotClean) {
+      }
+      DType colType = nestedChildren.getType();
+      Optional<Long> nullCount = nestedChildren.getNullCount();
+      //TODO check long to int is ok
```
Can we do this here or have a follow on issue for it?

revans2 commented on java/src/main/java/ai/rapids/cudf/ColumnVector.java:
```diff
@@ -2845,21 +3122,181 @@ protected boolean cleanImpl(boolean logErrorIfNotClean) {
+    }
+
+    private static long getEndStringOffset(long index, HostMemoryBuffer offsets) {
+      //TODO check for row count wrt to index
```
again either do this here or we need a follow on issue for it.

revans2 commented on java/src/main/java/ai/rapids/cudf/ColumnVector.java:
```diff
@@ -2845,21 +3122,181 @@ protected boolean cleanImpl(boolean logErrorIfNotClean) {
+      return offsets.getInt((index + 1) * 4);
+    }
+
+    static void printBuffer(HostMemoryBuffer buffer) {
```
If we want this for debugging then we should put it in HostMemoryBuffer instead.  Or else we should delete it.

revans2 commented on java/src/main/java/ai/rapids/cudf/ColumnVector.java:
```diff
@@ -2845,21 +3122,181 @@ protected boolean cleanImpl(boolean logErrorIfNotClean) {
+        if (type == DType.STRING) {
+          // This needs a different type
+          dataLen = getEndStringOffset(rows - 1, offsetBuffer);
+          //TODO check nullcount
```
Again do it or need it to be a part of a follow on issue.

revans2 commented on java/src/main/java/ai/rapids/cudf/ColumnVector.java:
```diff
@@ -2845,21 +3122,181 @@ protected boolean cleanImpl(boolean logErrorIfNotClean) {
+      }
+      System.out.println();
+    }
+    private static NestedColumnVector createNestedColumnVector(DType type, int rows, Optional<Long> nullCount,
```
nit: space above

revans2 commented on java/src/main/java/ai/rapids/cudf/ColumnVector.java:
```diff
@@ -2892,6 +3329,12 @@ public static ColumnVector boolFromBytes(byte... values) {
     return build(DType.BOOL8, values.length, (b) -> b.appendArray(values));
   }
 
+  public static<T> ColumnVector fromLists(HostColumnVector.ColumnBuilder.DataType dataType, List<T>... lists) {
```
Could we add in some java docs to say that this is unstable?  I just don't want people getting sued to this API if we think it is likely to change in later releases.

revans2 commented on java/src/main/java/ai/rapids/cudf/DeviceMemoryBuffer.java:
```diff
@@ -53,7 +53,7 @@ protected boolean cleanImpl(boolean logErrorIfNotClean) {
           // therefore it is unsafe to retry.
           address = 0;
-          lengthInBytes = 0;
+//          lengthInBytes = 0;
```
why?

revans2 commented on java/src/main/java/ai/rapids/cudf/HostColumnVector.java:
```diff
@@ -42,11 +47,137 @@
+      return nestedChildren;
+    }
+
+
```
nit: extra space

revans2 commented on java/src/main/java/ai/rapids/cudf/HostColumnVector.java:
```diff
@@ -87,7 +234,6 @@
     refCount = 0;
     incRefCountInternal(true);
   }
-
```
nit add back in the space here

kuhushukla replied:
done.

revans2 commented on java/src/main/java/ai/rapids/cudf/HostColumnVector.java:
```diff
@@ -200,7 +349,6 @@ public boolean hasValidityVector() {
   public boolean hasNulls() {
     return getNullCount() > 0;
   }
-
```
here too add back in the space.

kuhushukla replied:
done.

revans2 commented on java/src/main/java/ai/rapids/cudf/HostColumnVector.java:
```diff
@@ -528,6 +703,7 @@ public boolean isClean() {
     public long getHostMemorySize() {
       long total = 0;
       if (valid != null) {
+        //TODO: fix for lists
```
This needs to be done, but can be done in a follow on if necessary.  We should also make sure that it happens for the device side, which I think is only looking at the top level buffers too.

kuhushukla replied:
Did so something here that should account for children in lists. please take a look if it seems ok.

revans2 commented on java/src/main/java/ai/rapids/cudf/HostColumnVector.java:
```diff
@@ -594,6 +770,11 @@ public static HostColumnVector build(int rows, long stringBufferSize, Consumer<B
   }
 
+  public static<T> HostColumnVector fromLists(ColumnBuilder.DataType dataType, List<T>... values) {
+   ColumnBuilder cb = new ColumnBuilder(dataType);
```
nit: spacing appears off

kuhushukla replied:
Fixed

revans2 commented on java/src/main/java/ai/rapids/cudf/HostColumnVector.java:
```diff
@@ -981,19 +1162,418 @@ public static HostColumnVector timestampNanoSecondsFromBoxedLongs(Long... values
+  /**
+   * WARNING: Debug only method to print a passed in buffer
+   */
+  private static void printBuffer(HostMemoryBuffer buffer) {
```
I think this is a duplicate with the other one that should both probably be a part of HostMemoryBuffer.

kuhushukla replied:
I think now we have only one printBuffer method.

kuhushukla commented:
@revans2 , I think I have most of the comments from earlier reviews addressed. I am now rebasing and fixing some formatting. I have also moved this into REVIEW state. Please let me know if you find more concerns with the code and I will get to them as part of the next round of commits.

revans2 commented on java/src/main/java/ai/rapids/cudf/ColumnVector.java:
```diff
@@ -24,18 +24,91 @@
+    protected long viewHandle;
+
+    public DeviceColumnViewPointerAccess(long viewHandle) {
+      this.viewHandle = viewHandle;
```
This is a memory leak. In the C++ code the only way to get a view out is to copy it.  Which the native code does, but we don't have a close method on this that will free the host memory associated with the column view that we just copied out. I was hopeful that we could just reuse the same pointer that is stored in the `_children` vector, but that probably violates something with the abstraction for the C++ layer, so we are going to need to be sure that this gets closed each time it is used.

kuhushukla replied:
I have made some changes that should close the column view. Let me know what you think.

kuhushukla replied:
@revans2 Please see the use of .close() on the ColumnViewPointerAccess as that is how I am tackling it right now. Please correct  me if I missed something.

revans2 replied:
I would really prefer to have us use a try/finally where ever possible.  But this works.

revans2 commented on java/src/test/java/ai/rapids/cudf/ColumnVectorTest.java:
```diff
@@ -804,7 +809,7 @@ void testFromScalar() {
     final int rowCount = 4;
     for (DType type : DType.values()) {
-      if (type == DType.EMPTY) {
+      if (type == DType.EMPTY || type == DType.LIST) {
```
We need a follow on issue to support scalar lists.

kuhushukla replied:
Yes.

revans2 commented on java/src/main/java/ai/rapids/cudf/HostColumnVector.java:
```diff
@@ -981,19 +1186,419 @@ public static HostColumnVector timestampNanoSecondsFromBoxedLongs(Long... values
+
+    private void setNullAt(int index) {
+      initValidBuffer();
+      //TODO does this work
```
Sure looks like it does

revans2 commented on java/src/main/java/ai/rapids/cudf/HostColumnVector.java:
```diff
@@ -981,19 +1186,419 @@ public static HostColumnVector timestampNanoSecondsFromBoxedLongs(Long... values
+
+    @Override
+    public void close() throws Exception {
+      //TODO
```
I think we need to do something here.

revans2 commented on java/src/main/java/ai/rapids/cudf/HostColumnVector.java:
```diff
@@ -1369,7 +1974,7 @@ public final Builder append(HostColumnVector columnVector) {
     private void allocateBitmaskAndSetDefaultValues() {
       long bitmaskSize = ColumnVector.getNativeValidPointerSize((int) rows);
-      valid = HostMemoryBuffer.allocate(bitmaskSize);
+      valid = (HostMemoryBuffer.allocate(bitmaskSize));
```
nit: change is not needed

revans2 commented on java/src/main/java/ai/rapids/cudf/HostColumnVector.java:
```diff
@@ -1404,9 +2009,6 @@ public final Builder setNullAt(long index) {
     public final HostColumnVector build() {
-      if (built) {
-        throw new IllegalStateException("Cannot reuse a builder.");
-      }
```
Do we want to reuse a builder now?

kuhushukla replied:
I have for now made it so that u cannot. Maybe we will in the future..

revans2 commented on java/src/main/native/src/ColumnVectorJni.cpp:
```diff
@@ -1097,15 +1106,44 @@ JNIEXPORT jlong JNICALL Java_ai_rapids_cudf_ColumnVector_makeCudfColumnView(
+JNIEXPORT jlong JNICALL Java_ai_rapids_cudf_ColumnVector_getChildColumnView(JNIEnv *env,
+    jobject j_object, jlong handle) {
+    cudf::jni::auto_set_device(env);
+//    std::unique_ptr<cudf::column_view> ret;
```
Can we clean up the commented out code please?

revans2 commented on java/src/main/native/src/ColumnVectorJni.cpp:
```diff
@@ -1170,6 +1208,26 @@ JNIEXPORT jlongArray JNICALL Java_ai_rapids_cudf_ColumnVector_getNativeDataPoint
+                ret[0] = reinterpret_cast<jlong>(str_view.chars().data<char>());
+              ret[1] = str_view.chars().size();
+              } else {
+              ret[0] = reinterpret_cast<jlong>(data_view.data<char>());
```
indentation is off.

revans2 commented on java/src/main/native/src/ColumnVectorJni.cpp:
```diff
@@ -1195,7 +1280,17 @@ JNIEXPORT jlongArray JNICALL Java_ai_rapids_cudf_ColumnVector_getNativeOffsetsPo
+                   ret[0] = 0;
+                   ret[1] = 0;
+                 }
+               } else {
```
indentation appears to be off for all of this.

revans2 commented on java/src/test/java/ai/rapids/cudf/TableTest.java:
```diff
@@ -158,6 +158,14 @@ public static void assertPartialColumnsAreEqual(HostColumnVector expected, long
             break;
+          case LIST:
+            try {
+              assertArrayEquals(expected.getList(expectedRow).toString().getBytes(), cv.getList(tableRow).toString().getBytes(),
```
Can we have a follow on issue to fix this please?  We should not be doing string comparison for the lists.  This makes it so we cannot do approximate float comparisons or make sure that the types match or null count matches for a list of strings that has nulls in them....

kuhushukla replied:
Yes. agree this is not grt.

revans2 commented on java/src/test/java/ai/rapids/cudf/TableTest.java:
```diff
@@ -158,6 +158,14 @@ public static void assertPartialColumnsAreEqual(HostColumnVector expected, long
+              assertArrayEquals(expected.getList(expectedRow).toString().getBytes(), cv.getList(tableRow).toString().getBytes(),
+                  "Column " + colName + " Row " + tableRow);
+            } catch (Exception e) {
+              e.printStackTrace();
```
This should be removed too.

revans2 commented on java/src/main/java/ai/rapids/cudf/ColumnViewPointerAccess.java:
```diff
@@ -0,0 +1,45 @@
+/**
+ * Interface that allows access only methods to cudf column_views
+ */
+interface ColumnViewPointerAccess<T> {
```
In the spark plugin we need a way to get the child types, or else we are going to have to change a massive amount of code.  Can we please make this public?

kuhushukla commented:
> I still don't understand why data is different. I understand that we are treating offsets differently and so we have to play around with the number of children and offsets into the children, but data should just be in the leaf column right? If we play the same game that we did for Strings don't we risk losing the validity for that leaf column when we do this? If that is what we are doing here then I would argue that eventually we should just make string match list.
>

I am sorry @revans2 for this confusion. I am changing the way the jni call works now to better show what I want. To allow for getDataBuffer to somewhat work better with structs, I want this to be something where if pass a handle to getNativeData() it return  null if it is a complex type or actual buffer if not. That is what I am trying to capture here and eventually we will have an index passed in that will give us the individual child column view's data. If this is not something you agree with, I can revert this change or redo it in a way that fits better for structs.

kuhushukla commented on java/src/main/java/ai/rapids/cudf/ColumnVector.java:
```diff
@@ -257,6 +354,74 @@ public boolean hasNulls() {
+        ColumnViewPointerAccess childDevPtr = deviceCvPointer.getChildColumnView(i);
+        children.add(copyToHostNestedHelper(childDevPtr));
+        childDevPtr.close();
+      }
```
Since this would eventually throw a run time exception I just closed this without a try catch. Let me know if that doesnt agree with you, @revans2 .

kuhushukla replied:
@revans2 I have made some changes to the base interface and made it a try with resources block in all but one place where we are looping which makes the scopes harder. Let me know if I missed a nicer way to do this.

revans2 commented on java/src/main/java/ai/rapids/cudf/ColumnViewPointerAccess.java:
```diff
@@ -0,0 +1,47 @@
+
+  int getNumChildren();
+
+  void close();
```
Can we make this `AutoCloseable` instead?

kuhushukla replied:
Sure. I had it that way but I am curiosu what to do in copyToHost if we do have an exception? deleteCudfColumnView is throwing a runtime error so should be throw a cudf exception or have copyToHost throw it.

kuhushukla replied:
Having it be part of copyToHost signature makes for a lot of test code change.

kuhushukla replied:
This is addressed.

kuhushukla commented on java/src/main/java/ai/rapids/cudf/ColumnVector.java:
```diff
@@ -257,6 +357,74 @@ public boolean hasNulls() {
+        hostOffsets = HostMemoryBuffer.allocate(currOffsets.length);
+        hostOffsets.copyFromDeviceBuffer(currOffsets);
+      }
+      for (int i = 0; i < deviceCvPointer.getNumChildren(); i++) {
```
done.

kuhushukla commented on java/src/main/java/ai/rapids/cudf/ColumnVector.java:
```diff
@@ -24,18 +24,100 @@
+
+    /**
+     * IMPORTANT: This returns a new Column View, so you MUST close it once you are done with this.
+     */
```
done.

kuhushukla commented:
> would really prefer to see us use a try with close block which should be simple if you make ColumnViewPointerAcesss an AutoCloseable
>
@revans2 I have made some changes to the base interface and made it a try with resources block in all but one place where we are looping which makes the scopes harder. Let me know if I missed a nicer way to do this.

revans2 commented:
@mythrocks and @razajafri could you both take another look?

razajafri commented on java/src/main/java/ai/rapids/cudf/ColumnVector.java:
```diff
@@ -24,18 +24,99 @@
+    }
+
+    @Override
+    public long getColumnView() {
```
I am wondering if this method should be renamed to something that conveys what its returning. may be `getColumnViewAddress`??

kuhushukla replied:
That wont be like `getNativeView` anymore then. Let me know what u think of that.

razajafri replied:
That is a private method. I am mostly concerned about the public API calling it a view when its a long.

kuhushukla replied:
u got it! I will rename this.

razajafri commented on java/src/main/java/ai/rapids/cudf/ColumnVector.java:
```diff
@@ -24,18 +24,99 @@
+    }
+
+    @Override
+    public ColumnViewPointerAccess<BaseDeviceMemoryBuffer> getChildColumnView(int childIndex) {
```
This should be called `getChilColumnViewPointerAccess` exactly what it returns

kuhushukla replied:
:( very long name :( but happy to make the change.

kuhushukla replied:
The more I think about it - it is not a simple getter on the class so maybe we can keep this one as is , what do u think @razajafri ?

razajafri replied:
Will it make sense to change the name of the class to be ColumnViewAccess? That will make it shorter and avoid the ambiguity

kuhushukla replied:
sounds good. @revans2 I know your related PRs will be impacted by this public api change, let me know if that's cool with u and I will psuh a change here soon.

kuhushukla replied:
Made the change for now. @razajafri please review :)

razajafri commented on java/src/main/java/ai/rapids/cudf/ColumnVector.java:
```diff
@@ -24,18 +24,99 @@
+
+    @Override
+    public int getNumChildren() {
+      if (getDataType() != DType.LIST /* or structs */) {
```
what about strings? it has two children I believe

kuhushukla replied:
Here, this method, if u look at the jni call as well, gives u number of `child`s of a column view.(excluding offsets since we dont quite need them thru this call)
strings - number of child columns = 0
list - number of child columns = 1
struct - number of child columns  >= 1

mythrocks replied:
> struct - number of child columns >= 1

`>=0`. :]

kuhushukla replied:
Sure :). That's true with any column type really. I was trying to demonstrate what this could return.  num_children should indicate that isn't it?

kuhushukla commented:
>Assuming that we have follow on issues to address

  1.  Updating the tests to not use strings, but to actually recurse into the child columns.
  2.   List Scalar support.
>
Issues https://github.com/rapidsai/cudf/issues/5888 and https://github.com/rapidsai/cudf/issues/5887 for 1. and 2. respectively.

> I would also like us to have a follow on issue to look at how to make more user friendly ways of building the test data.
>
https://github.com/rapidsai/cudf/issues/5889 for this one.

kuhushukla commented on java/src/main/native/src/ColumnVectorJni.cpp:
```diff
@@ -1262,6 +1281,40 @@ JNIEXPORT jlongArray JNICALL Java_ai_rapids_cudf_ColumnVector_getNativeDataPoint
+      if (column->type().id() == cudf::type_id::STRING) {
+        return 0;
+      }
+      // first child is always offsets which we do not want to count here
```
@razajafri making doubly sure we cover strings although the java side wont let u come here per current change/design. Please review.

razajafri commented on java/src/main/java/ai/rapids/cudf/ColumnVector.java:
```diff
@@ -24,18 +24,99 @@
+
+    @Override
+    public ColumnViewAccess<BaseDeviceMemoryBuffer> getChildColumnViewAccess(int childIndex) {
+      if (getDataType() != DType.LIST) {
```
check if the child index is within bounds by `getNumChildren`

kuhushukla replied:
Fixed.

razajafri commented on java/src/main/java/ai/rapids/cudf/HostColumnVector.java:
```diff
@@ -42,11 +44,152 @@
 
+  protected static class NestedHostColumnVector implements AutoCloseable {
+
+    private HostMemoryBuffer data;
+    private HostMemoryBuffer offsets;
+    private HostMemoryBuffer validity;
```
Why can't this just be an offHeapState? 

For that matter, correct me if I am wrong, I was thinking the only reason we have to create this NestedHostColumnVector class is because of refcount. Can we extract a common class between the HostColumnVector and the NestedHostColumnVector? Then the HostColumnVector can just extend it and add refCount.

kuhushukla replied:
This is similar to Bobby's review comment from earlier 
```
Another nit for follow on work. After refactoring the types being returned we probably could combine the majority of this code with the code from copyToHostNestedHelper so that all this does is call into copyToHostNestedHelper and then has an API to translate the top level HostColumnVector.NestedHostColumnVector into a HostColumnVector. But again lets get this in as is and worry about that later.
```
@rjafri , agree this can be done. I relaized it during the review process and can be done incrementally/next version. Let me know if you are ok with a  follow on. Thanks!

razajafri replied:
Thank you @kuhushukla. I think a follow-on should be good

kuhushukla replied:
https://github.com/rapidsai/cudf/issues/5905 for this issue and the readValue minor comment on consolidation for strings

razajafri commented on java/src/main/java/ai/rapids/cudf/HostColumnVector.java:
```diff
@@ -42,11 +44,152 @@
+        case INT8: return data.getByte(index);
+        case INT16: return data.getShort(index);
+        case BOOL8: return data.getBoolean(index);
+        default: throw new UnsupportedOperationException("Do not support " + type);
```
Are Strings not supported by lists?

kuhushukla replied:
Strings are Supported per code in `getElement` and I have added tests as well to show how List of strings work. Let me know if you have any questions.

kuhushukla commented on java/src/main/java/ai/rapids/cudf/ColumnVector.java:
```diff
@@ -265,33 +433,86 @@ public HostColumnVector copyToHost() {
       HostMemoryBuffer hostDataBuffer = null;
```
This is not part of this change as it was present from earlier changes. Let me know if that works, @razajafri .

razajafri replied:
I know its not a part of your change. The change is too small of a nit to warrant its own PR. Its not a big deal I don't want to hold the PR for this, but if you make a change it will be nice to see this

kuhushukla commented on java/src/main/java/ai/rapids/cudf/ColumnVector.java:
```diff
@@ -265,33 +433,86 @@ public HostColumnVector copyToHost() {
+            hValid = HostMemoryBuffer.allocate(valid.getLength());
+            hValid.copyFromDeviceBuffer(valid);
+          }
+            if (offsets != null) {
```
Fixed.

kuhushukla commented on java/src/main/java/ai/rapids/cudf/ColumnVector.java:
```diff
@@ -265,33 +433,86 @@ public HostColumnVector copyToHost() {
+              hOffset = HostMemoryBuffer.allocate(offsets.getLength());
+              hOffset.copyFromDeviceBuffer(offsets);
+          }
+            List<HostColumnVector.NestedHostColumnVector> children = new ArrayList<>();
+            try(ColumnViewAccess childDevPtr = getChildColumnViewAccess(0)) {
+              children.add(copyToHostNestedHelper(childDevPtr));
+            }
```
Fixed.

kuhushukla commented:
@razajafri , I have addressed all the comments that needed a code change and left responses for the ones with clarification. Please take a look. Thanks a lot!
